src/ann_solo/utils.py

```diff
@@ -1,104 +1,338 @@
+            features["precursor_charge_2"].append(1)
+            features["precursor_charge_3"].append(0)
+            features["precursor_charge_4"].append(0)
+            features["precursor_charge_5"].append(0)
+        elif ssm.query_spectrum.precursor_charge == 3:
+            features["precursor_charge_2"].append(0)
+            features["precursor_charge_3"].append(1)
+            features["precursor_charge_4"].append(0)
+            features["precursor_charge_5"].append(0)
+        elif ssm.query_spectrum.precursor_charge == 4:
+            features["precursor_charge_2"].append(0)
+            features["precursor_charge_3"].append(0)
+            features["precursor_charge_4"].append(1)
+            features["precursor_charge_5"].append(0)
+        elif ssm.query_spectrum.precursor_charge >= 5:
+            features["precursor_charge_2"].append(0)
+            features["precursor_charge_3"].append(0)
+            features["precursor_charge_4"].append(0)
+            features["precursor_charge_5"].append(1)
+        features["query_prec_mz"].append(ssm.query_spectrum.precursor_mz)
+        features["lib_prec_mz"].append(ssm.library_spectrum.precursor_mz)
+        features["mz_diff_ppm"].append(
+            mass_diff(
+                ssm.query_spectrum.precursor_mz,
+                ssm.library_spectrum.precursor_mz,
+                False,
+            )
+        )
+        features["cosine"].append(sim.cosine(ssm))
+        features["frac_n_peaks_query"].append(sim.frac_n_peaks_query(ssm))
+        features["frac_n_peaks_lib"].append(sim.frac_n_peaks_library(ssm))
+        features["frac_int_query"].append(sim.frac_intensity_query(ssm))
+        features["frac_int_lib"].append(sim.frac_intensity_library(ssm))
+        features["mse_mz"].append(sim.mean_squared_error(ssm, "mz"))
+        features["mse_int"].append(sim.mean_squared_error(ssm, "intensity"))
+        features["contrast_angle"].append(sim.spectral_contrast_angle(ssm))
+        features["hypergeometric_score"].append(sim.hypergeometric_score(ssm))
+        features["kendalltau"].append(sim.kendalltau(ssm))
+        features["ms_for_id_v1"].append(sim.ms_for_id_v1(ssm))
+        features["ms_for_id_v2"].append(sim.ms_for_id_v2(ssm))
+        features["entropy_unweighted"].append(sim.entropy(ssm, False))
+        features["entropy_weighted"].append(sim.entropy(ssm, True))
+        features["scribe_fragment_acc"].append(sim.scribe_fragment_acc(ssm))
+        features["manhattan"].append(sim.manhattan(ssm))
+        features["euclidean"].append(sim.euclidean(ssm))
+        features["chebyshev"].append(sim.chebyshev(ssm))
+        features["pearsonr"].append(sim.pearsonr(ssm))
+        features["braycurtis"].append(sim.braycurtis(ssm))
+        features["canberra"].append(sim.canberra(ssm))
+        features["ruzicka"].append(sim.ruzicka(ssm))
+        features["is_target"].append(not ssm.is_decoy)
+        features["group"].append("G")
```

Comment:
I don't understand this change. This would assign everything to the same group, rather than the groups per modification that we want.

Reply:
In the code you changed previously, you assign the group from the class SpectrumSpectrumMatch which doesn't even exist as an attribute, while executing the whole code was getting an error here. Second, this line of code is kinda useless as we call _get_ssm_groups() in the next statement (Line 74) which assigns new groups for this column. This line (337) can even be omitted.

Reply:
Oh, that's right. Must have been some incomplete changes on my part. Yes, better to just remove this line here then, rather than have it be confusing.